src/java.net.http/share/classes/jdk/internal/net/http/RequestPublishers.java

```diff
@@ -418,6 +419,90 @@ public long contentLength() {
         }
     }
 
+    public static final class FileChannelPublisher implements BodyPublisher {
+
+        private final FileChannel channel;
+
+        private final long position;
+
+        private final long limit;
+
+        public FileChannelPublisher(FileChannel channel, long offset, long length) {
+            this.channel = Objects.requireNonNull(channel, "channel");
+            long fileSize = fileSize(channel);
+            Objects.checkFromIndexSize(offset, length, fileSize);
+            this.position = offset;
+            this.limit = offset + length;
+        }
+
+        private static long fileSize(FileChannel channel) {
+            try {
+                return channel.size();
+            } catch (IOException ioe) {
+                throw new UncheckedIOException(ioe);
+            }
+        }
+
+        @Override
+        public long contentLength() {
+            return limit - position;
+        }
+
+        @Override
+        public void subscribe(Flow.Subscriber<? super ByteBuffer> subscriber) {
+            Iterable<ByteBuffer> iterable = () -> new FileChannelIterator(channel, position, limit);
+            new PullPublisher<>(iterable).subscribe(subscriber);
+        }
+
+    }
+
+    private static final class FileChannelIterator implements Iterator<ByteBuffer> {
+
+        private final FileChannel channel;
+
+        private final long limit;
+
+        private long position;
+
+        private boolean terminated;
+
+        private FileChannelIterator(FileChannel channel, long position, long limit) {
+            this.channel = channel;
+            this.position = position;
+            this.limit = limit;
+        }
+
+        @Override
+        public synchronized boolean hasNext() {
+            return position < limit && !terminated;
+        }
+
+        @Override
+        public synchronized ByteBuffer next() {
+            if (!hasNext()) {
+                throw new NoSuchElementException();
+            }
+            long remaining = limit - position;
+            ByteBuffer buffer = Utils.getBufferWithAtMost(remaining);
+            try {
+                int readLength = channel.read(buffer, position);
+                // Short-circuit if `read()` has failed, e.g., due to file content being changed in the meantime
+                if (readLength < 0) {
```

Comment:
The `FileChannel.read(ByteBuffer, position)` API states:

> If the given position is greater than or equal to the file's current size then no bytes are read.

I think we might have to consider the situation where the underlying file's size has reduced and this position is now past that size and the `read()` call returning 0. Would that then end up in a situation where we have a "never ending" BodyPublisher, which publishes nothing but hasn't terminated either?

Reply:
The `FileChannel.read(ByteBuffer, position)` API further states the following:

> [Method returns] The number of bytes read, possibly zero, or `-1` if the given position is greater than or equal to the file's current size

That is, the case you're describing is indicated by the `-1` return value, which we check in the next line. Plus, we already have a test for precisely the scenario that you're describing: `testFileModificationDuringPublisherRead`.